service/matching/taskqueuemonitor.go

```
@@ -0,0 +1,116 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package matching
+
+import (
+	"context"
+	"math"
+	"time"
+)
+
+type (
+	tqmonStop    struct{}
+	tqmonGC      struct{}
+	tqmonUsurped struct{ id *taskQueueID }
+
+	// tqmonAccess is the interface by which a taskQueueMonitor observes and
+	// updates a set of taskQueueManager instances.
+	tqmonAccess interface {
+		getTaskQueues(int) []taskQueueManager
+		unloadTaskQueue(*taskQueueID)
+	}
+
+	// taskQueueMonitor performs garbage collection and unloads instances of
+	// taskQueueManager accessed via a tqmonAccess.
+	taskQueueMonitor struct {
+		sigs chan interface{}
+	}
+)
+
+// newTaskQueueMonitor constructs a new instance of taskQueueMonitor.
+func newTaskQueueMonitor() *taskQueueMonitor {
+	return &taskQueueMonitor{sigs: make(chan interface{})}
+}
+
+// Run runs the taskQueueMonitor event loop, performing GC of a set of
+// taskQueueManager instances and unloading the same when signaled or when GC
+// deems it appropriate. The implementaion of tqmonAccess MUST NOT re-enter this
+// taskQueueMonitor instance.
+func (m *taskQueueMonitor) Run(a tqmonAccess, c *Config) {
+	defer close(m.sigs)
+	t := time.NewTimer(c.IdleTaskqueueCheckInterval())
+	defer t.Stop()
+
+	for {
+		select {
+		case i := <-m.sigs:
+			switch sig := i.(type) {
+			case tqmonUsurped:
+				a.unloadTaskQueue(sig.id)
+			case tqmonGC:
+				m.runGC(a)
+				t.Reset(c.IdleTaskqueueCheckInterval())
+			case tqmonStop:
+				return
+			}
```

can you add a default here? in case future impl add more type
the default case can simply panic

---

It was intentional that I left it out. If someone throws a `string` into the signal channel I figured it was better to just ignore it than to panic. But I can crash the server.

---

it will be preferable to just crash (fail fast)